### PR TITLE
restore: elision of scattering when max-replica -ge store-count

### LINKED ISSUE #27425: restore database blocked
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1 tidb, 1 pd, 1 tikv
```
use test;
// create many list partition tables;
BACKUP DATABASE test to 'local:///tmp/lp_backup/';
drop database test;
create database test;
use test;
RESTORE DATABASE * FROM 'local:///tmp/lp_backup/';
```

### 2. What did you expect to see? (Required)
restore success

### 3. What did you see instead (Required)
```
mysql> show processlist;
+------+------+----------------------+------+---------+-------+------------+---------------------------------------------------+
| Id   | User | Host                 | db   | Command | Time  | State      | Info                                              |
+------+------+----------------------+------+---------+-------+------------+---------------------------------------------------+
|   49 | root | 192.168.228.89:59988 | test | Query   | 12070 | autocommit | RESTORE DATABASE * FROM 'local:///tmp/lp_backup/' |
|   57 | root | 192.168.228.89:54784 | NULL | Query   |     0 | autocommit | show processlist                                  |
+------+------+----------------------+------+---------+-------+------------+---------------------------------------------------+
2 rows in set (0.00 sec)
```
### 4. What is your TiDB version? (Required)

```
Release Version: v5.2.0
Edition: Community
Git Commit Hash: d4d667fba5e82f108c62cd690fcf73c685d7612f
Git Branch: heads/refs/tags/v5.2.0
UTC Build Time: 2021-08-18 12:35:12
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #27425 

Problem Summary: After retrying failed scatter requests, when `max-replica > store-number`, it would always fail. A full retry costs about 10+ seconds, which is unacceptable for upon thousands even millions of regions. 

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:
When `max-replica >= store-number`, skip scattering, since it is no meaning to scatter in this condition. 

How it Works:
So, the retry never happen.

### Check List 

Tests 

- [ ] Manual test (add detailed scripts or steps below)

In a typical TiUP playground (store-count = 1, max-replica = 3):
image
In a testing cluster (store-count = 3, max-replica = 1):
image

Side effects

- [ ] Performance regression: Consumes more Networking (O(1))

### Release note 

```release-note
Fix a bug that causes BR get stuck in some of small clusters.
```

### CODE REVIEW

**Comment by kennytm** [comment]
when `max-replica == store-count > 1`, without scattering there will be a large leader imbalance (though size is balanced). not sure if PD can quickly fix it afterwards.

maybe just skip scatter when `max-replica > store-count || store-count = 1`?

File: br/pkg/restore/split_client.go [hunk 1]
```diff
@@ -374,7 +397,56 @@ func (c *pdClient) BatchSplitRegions(
...
+	}
+	log.Info("checking whether need to scatter", zap.Int("store", storeCount), zap.Int("max-replica", maxReplica))
+	// Skipping scatter may lead to leader unbalanced,
+	// currently, we skip scatter only when:
+	//   1. max-replica > store-count (Probably a misconfigured or playground cluster.)
+	//   2. store-count == 1 (No meaning for scattering.)
+	// We can still omit scatter when `max-replica == store-count`, if we create a BalanceLeader operator here,
+	//   however, there isn't evidence for transform leader is much faster than scattering empty regions.
+	doNotNeedScatter := maxReplica > storeCount || storeCount == 1
+	return !doNotNeedScatter, nil
```
Comment (kennytm) [inline br/pkg/restore/split_client.go:443]:
```suggestion
	return storeCount >= maxReplica && storeCount > 1, nil
```

**Review by kennytm — APPROVED** [review]
rest LGTM

**Review by 3pointer — APPROVED** [review]

**Pull request merged**